Feature/choice set manager

crokita opened the pull request:
Partially fixes #123 

### Risk
This PR makes minor API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Unit tests added that match the tests added in the Java Suite library for the choice set related classes

### Summary
Adds the choice set manager which feature matches the Java Suite library for all the screen manager APIs related to the choice set manager.

Note that presentSearchableChoiceSet would not be supported in the latest develop Generic HMI, and that OnKeyboardInput messages haven't been received.

renonick87 commented on lib/js/src/manager/screen/choiceset/ChoiceCell.js:
```diff
@@ -0,0 +1,252 @@
+        if (this.getArtwork() === null && other.getArtwork() !== null) {
+            return false;
+        }
+        if (this.getArtwork() !== null && !this.getArtwork().equals(other.getArtwork())) {
```
The .equals() method is part of the SdlFile class, not the SdlArtwork class. Because of that, it won't check the _isTemplate property which is part of the SdlArtwork class.

crokita replied:
I believe this is the way the Java version does this. Their way of comparing object equality is by implementing the `hashCode`/`equals` method. The SdlArtwork class doesn't implement this but the parent SdlFile does. Unless this is a bug with the Java code as well, it seems to ignore the extra properties SdlArtwork offers as part of the equality check.

renonick87 commented on lib/js/src/manager/screen/choiceset/_CheckChoiceVrOptionalOperation.js:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+
+        const deleteResponse = await deletePromise;
```
You're using await here but that was already done when deletePromise was initialized earlier.

renonick87 commented on lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js:
```diff
@@ -0,0 +1,525 @@
+        const choicesToBeRemovedFromPending = this._keepChoicesInBoth(choices, this._pendingPreloadChoices);
+        // If choices are deleted that are already uploaded or pending and are used by a pending presentation, cancel it and send an error
+        let pendingPresentationChoices = [];
+        if (this._pendingPresentationSet !== null) { // type is ChoiceSet
```
```suggestion
        if (this._pendingPresentationSet !== null && this._pendingPresentationSet.getChoices() !== null) { // type is ChoiceSet
```

renonick87 commented on lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js:
```diff
@@ -0,0 +1,262 @@
+        this._initialText = initialText;
+        this._customConfig = customConfig;
+        this._keyboardListener = keyboardListener;
+        this._cancelId = cancelId;
```
```suggestion
        this._cancelId = cancelId;
        this._keyboardProperties = originalKeyboardProperties;
```

renonick87 commented on lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js:
```diff
@@ -0,0 +1,262 @@
+                .setCancelID(this._cancelId);
+
+            const response = await this._lifecycleManager.sendRpcResolve(cancelInteraction);
+            if (!response.getSuccess()) {
+                console.error('Canceled the keyboard set unsuccessfully');
+            }
```
```suggestion
            console.error(`Canceled the keyboard set ${response.getSuccess() === Result.SUCCESS ? 'successfully' : 'unsuccessfully'}`);
```

renonick87 reviewed (changes requested):
See requested feedback and remember to run `npm run lint` to resolve any issues in documentation.

crokita commented:
PR updated. I didn't include unrelated linter fixes in this PR: that can be done in a separate PR. Also note:
- I moved systemcapabilitylistener logic to the _ChoiceSetManagerBase to fix an issue with _displayName never being set by looking at the display capabilities
- _ChoiceSetSelectionListener and _KeyboardListener are now public classes
- Fixed ChoiceCell's equals() method to not error out when comparing null voice commands

renonick87 commented on lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js:
```diff
@@ -508,14 +516,46 @@ class _ChoiceSetManagerBase extends _SubManagerBase {
+     * Listen for DISPLAYS capability updates
+     * @private
+     */
+    _addListeners () {
```
This method looks to be the same as the `_SubManagerBase._handleDisplayCapabilities` method.

crokita replied:
Line 538 has the small difference. The display name from the system capabilities needs to be taken and saved as displayName in this class

renonick87 commented on tests/managers/index.js:
```diff
@@ -32,10 +40,19 @@ describe('ManagerTests', function () {
                 voiceCommandUpdateOperationTests(appClient);
```
Something with the most recent update is causing failures in the `voiceCommandManagerTests` and `voiceCommandUpdateOperationTests` for me. I've tested this with manticore across multiple browsers. Everything passes if these tests are moved above the `textAndGraphicManagerTests` and run first though.

crokita replied:
That's annoying. I have moved those tests and got the same results now.

crokita commented:
Updated. See comments in response to feedback

renonick87 reviewed (approved):
Since this PR contains updates from https://github.com/smartdevicelink/sdl_javascript_suite/pull/357, merging the ChoiceSetManager will have to wait until after that PR is merged first.

renonick87 commented:
Updates from https://github.com/smartdevicelink/sdl_javascript_suite/pull/357 have been reverted from this PR. This PR is set to be merged.

crokita commented:
Squashing to avoid accidentally clearing out this PR: #357
